text/3371-cargo-target-directories.md

```diff
@@ -0,0 +1,328 @@
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+<!--
+Explain the proposal as if it was already included in the language and you were teaching it to another Rust programmer. That generally means:
+
+- Introducing new named concepts.
+- Explaining the feature largely in terms of examples.
+- Explaining how Rust programmers should *think* about the feature, and how it should impact the way they use Rust. It should explain the impact as concretely as possible.
+- If applicable, provide sample error messages, deprecation warnings, or migration guidance.
+- If applicable, describe the differences between teaching this to existing Rust programmers and new Rust programmers.
+- Discuss how this impacts the ability to read, understand, and maintain Rust code. Code is read and modified far more often than written; will the proposed feature make code easier to maintain?
+
+For implementation-oriented RFCs (e.g. for compiler internals), this section should focus on how compiler contributors should think about the change, and give examples of its concrete impact. For policy RFCs, this section should provide an example-driven introduction to the policy, and explain its impact in concrete terms.
+-->
+
+For a single project, it is possible to use the `CARGO_TARGET_DIR` environment variable (or the `target-dir` TOML config option or the `--target-dir` command-line flag) to change the position of the `target/` directory used for build artifacts during compilation with Cargo.
+
+While this option is useful for single-project environments (simple CI builds, builds through other build systems like Meson or Bazel), in multi-projects environment, like personal machines or repos with multiple workspaces, it conflates every build directory under the configured path: `CARGO_TARGET_DIR` directly replaces the `<workspace>/target/` directory.
+
+`CARGO_TARGET_DIRECTORIES` (or the `target-directories` TOML option or the `--target-directories` command-line flag) instead acts as a parent for those `target` directories.
+
+Below is an example of the behavior with `CARGO_TARGET_DIR` versus the one with `CARGO_TARGET_DIRECTORIES`:
+
+## Example
+
+Consider this directory tree:
+
+```text
+/Users/
+├─ poliorcetics/
+│  ├─ work/
+│  │  ├─ work-project/
+│  │  │  ├─ Cargo.toml
+│  │  │  ├─ crate-1/
+│  │  │  │  ├─ Cargo.toml
+│  │  │  ├─ crate-2/
+│  │  │  │  ├─ Cargo.toml
+│  ├─ perso/
+│  │  ├─ perso-1/
+│  │  │  ├─ Cargo.toml
+│  │  ├─ perso-2/
+│  │  │  ├─ Cargo.toml
+
+/cargo-cache/
+```
+
+#### With `CARGO_TARGET_DIR=/cargo-cache`
+
+`cd /Users/poliorcetics/work/work-project && cargo build` produces artifacts directly in `/cargo-cache/debug/...`
+
+A subsequent `cargo build` in `project-1` will work with the same artifact, potentially having conflicting features for dependencies for example.
+
+A `cargo clean` will delete the entire `/cargo-cache` directory, for all projects at once.
```

Comment:
Doesn't this contradict your statement later

> Currently, if CARGO_TARGET_DIR is set to anything but target for a project, cargo clean does not delete the target/ directory if it exists. The same behavior is used for CARGO_TARGET_DIRECTORIES.

Reply:
Oh, I think you mean that it respects `CARGO_TARGET_DIR` and will not clean up anything else, like a `target/` directory?

Reply:
Yes, I clarified it